370-auth-ui/sdk.md

```diff
@@ -0,0 +1,306 @@
+interface OAuthError {
+  error: string;
+  error_description?: string;
+  error_uri?: string;
+}
+
+// The promise rejects with an JavaScript Error that looks like a OAuthError.
+function authorize(options: AuthorizeOptions): Promise<User>;
+
+// Open the URL with the user agent that is used to perform authentication.
+// On iOS it is SFAuthenticationSession or ASWebAuthenticationSession
+// On Android it is Chrome Custom Tabs.
+function openURL(url: string): Promise<void>;
+```
+
+#### Example
+
+##### Authenticate in a React Native application
+
+```typescript
+import React, { useCallback } from "react";
+import skygear from "@skygear/react-native";
+import {Button, View} from "react-native";
+function AuthenticationScreen({ navigate } : { navigate: (screenName: string) => void }) {
+  const onPress = useCallback(() => {
+    try {
+      const user = await skygear.auth.authorizeWithWeb({
+        redirectURI: "myappid://anyhost/anypath",
+        endpoint: "https://accounts.myapp.com",
+      });
+      // If the application requires verification, check if the user is verified.
+      if (!user.isVerified) {
+        navigate("VerificationScreen");
+      } else {
+        navigate("HomeScreen");
+      }
+    } catch (e) {
+      // Handle the error properly here.
+      // For example, duplicated user.
+    }
+  }, []);
+  return (
+    <View>
+      <Button title="Sign in" onPress={onPress} />
+    </View>
+  );
+}
+```
+
+##### Open settings in a React Native application
+
+```typescript
+import React, { useCallback } from "react";
+import skygear from "@skygear/react-native";
+import {Button, View} from "react-native";
+function SettingsScreen() {
+  const onPressEditProfile = useCallback(() => {
+    // Application-specific settings
+  }, []);
+  const onPressSecuritySettings = useCallback(() => {
+    skygear.auth.openURL("https://accounts.myapp.com").catch(e => {
+      // The error here should be unrecoverable.
+    });
+  }, []);
+  return (
+    <View>
+      <Button title="Edit Profile" onPress={onPressEditProfile} />
+      <Button title="Security Settings" onPress={onPressSecuritySettings} />
+    </View>
+  );
+}
+```
+
+### Browser-based App SDK
+
+#### API
+
+```typescript
+interface AuthorizeOptions {
+  redirectURI: string;
+  state?: string;
+  onUserDuplicate?: "abort" | "merge" | "create";
+}
+
+// Modeled after https://tools.ietf.org/html/rfc6749#section-5.2
+interface OAuthError {
+  error: string;
+  error_description?: string;
+  error_uri?: string;
+}
+
+// The promise resolves to nothing. It is because after calling this function,
+// the URL is changed. The app is destoryed so the return value is not usable at all.
+function authorize(options: AuthorizeOptions): Promise<void>;
+
+// exchangeToken looks at window.location
+// It checks if error is present and rejects with OAuthError.
+// Otherwise if code is present, make a token request.
+function exchangeToken(): Promise<User>;
```

Comment:
`state` is always there.

`state=state&code=code` or `state=state&error=invalid_request`.

If we only return `state` for the success case, it is a bit weird.

Reply:
Yes, we can return both. I mean `exchangeToken` would returns `{ user: User, state: unknown}`, or throws `class OAuthError { message: string; state: unknown; }`, so that developer don't need to extract the state from the URI manually.